jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -333,109 +333,116 @@ public ByteBuffer onUpgradeFrom()
     }
 
-    void releaseRequestBuffer()
+    private void releaseRequestBuffer()
     {
-        if (_retainableByteBuffer != null && !_retainableByteBuffer.hasRemaining())
-        {
-            if (LOG.isDebugEnabled())
-                LOG.debug("releaseRequestBuffer {}", this);
-            RetainableByteBuffer buffer = _retainableByteBuffer;
-            _retainableByteBuffer = null;
-            if (!buffer.release())
-                throw new IllegalStateException("unreleased buffer " + buffer);
-        }
+        if (LOG.isDebugEnabled())
+            LOG.debug("releasing request buffer {} {}", this, _requestBuffer);
+        _requestBuffer.release();
+        _requestBuffer = null;
     }
 
-    private ByteBuffer getRequestBuffer()
+    private void ensureRequestBuffer()
     {
-        if (_retainableByteBuffer == null)
-            _retainableByteBuffer = _bufferPool.acquire(getInputBufferSize(), isUseInputDirectByteBuffers());
-        return _retainableByteBuffer.getByteBuffer();
+        if (_requestBuffer == null)
+            _requestBuffer = _bufferPool.acquire(getInputBufferSize(), isUseInputDirectByteBuffers());
     }
 
     public boolean isRequestBufferEmpty()
     {
-        return _retainableByteBuffer == null || !_retainableByteBuffer.hasRemaining();
+        return _requestBuffer == null || !_requestBuffer.hasRemaining();
     }
 
     @Override
     public void onFillable()
     {
         if (LOG.isDebugEnabled())
-            LOG.debug(">>onFillable enter {} {} {}", this, _httpChannel, _retainableByteBuffer);
+            LOG.debug("onFillable enter {} {} {}", this, _httpChannel, _requestBuffer);
 
         HttpConnection last = setCurrentConnection(this);
         try
         {
+            ensureRequestBuffer();
+
             // We must loop until we fill -1 or there is an async pause in handling.
             // Note that the endpoint might already be closed in some special circumstances.
             while (true)
             {
-                // Fill the request buffer (if needed).
                 int filled = fillRequestBuffer();
                 if (LOG.isDebugEnabled())
-                    LOG.debug("onFillable filled {} {} {} {}", filled, this, _httpChannel, _retainableByteBuffer);
+                    LOG.debug("onFillable filled {} {} {} {}", filled, this, _httpChannel, _requestBuffer);
 
                 if (filled < 0 && getEndPoint().isOutputShutdown())
                     close();
 
-                // Parse the request buffer.
                 boolean handle = parseRequestBuffer();
 
                 // There could be a connection upgrade before handling
                 // the HTTP/1.1 request, for example PRI * HTTP/2.
                 // If there was a connection upgrade, the other
                 // connection took over, nothing more to do here.
                 if (getEndPoint().getConnection() != this)
+                {
+                    releaseRequestBuffer();
```

Comment:
I don't think you can release here, as we will loose bytes that would be transferred in a later `doUpgradeFrom`